### DISCUSSION TITLE
Is it possible to run x86_64 and aarch64 containers at the same time?

### DISCUSSION THREAD

**Discussion by wzxu** [abiosoft/colima#631] [Q&A] [answered]
On an Apple Silicon Mac, since Oracle can only run in x86_64 mode, I have no choice but to do so. However, the application I try to run is just too slow under emulation, so I'd like to run it in native aarch64 mode instead, and have it connect to the Oracle container in x86_64 mode through an exposed port. Is this possible?

My first assumption is not because it seems like there can only be one active Docker context at one time. Granted, I'm still not that familiar with Docker, and especially not with Colima, so maybe I'm wrong and there's a chance. Kindly help with an answer or some advice how to best handle said scenario. Thanks!

**Comment by rfay** [comment]
@wzxu are you having actual success running your Oracle images in x68_64 mode? Is it reliable? Is the performance even slightly tolerable? I'm interested because I always just say "no" to people who want to run amd64 images on arm64 with Colima or Docker Desktop.

**Reply by wzxu** [reply]
Yes it's a prebuilt image based on Oracle XE 18. It won't run in arm64 mode because something always goes wrong while starting the database. Gave it 4GB of RAM in x86_64 mode and it seems to run normally so far. Can't speak for performance since I haven't started to really work with it, but only created the schema and seeded it.

**Reply by rfay** [reply]
Thanks. If you "haven't really started to work with it" I'll be interested in future reports. Mostly it seems people who have tried this sort of thing run into unpredictable results.

**Reply by rayr007** [reply]
The performance of running Oracle XE using the image from container-registry.oracle.com/database/express:21.3.0-xe in colima with an x86_64 VM on my M1 MBP is better than it was running natively on my old 12" Macbook which had a core M5 processor. Actually, it is way faster; but that was a super low bar. 🤣

**Reply by wzxu** [reply]
@rfay I posted a quick update: https://github.com/abiosoft/colima/discussions/631#discussioncomment-5497404

**Comment by abiosoft** [comment, marked as answer]
You can run both at the same time if

- Colima is running with the host's architecture. i.e. no x86_64 emulation on m1
- You specify the `--platform` flag to `docker run` for cross architecture images.

Judging from your explanation, it appears you need to run Colima in emulation mode and that would not work. Except you would have multiple Colima profiles (and docker contexts) and toggle between them.


```sh
colima start # native
colima start --arch x86_64 intel # emulated

docker context use colima # switch to native context
docker context use colima-intel # switch to emulated

# you can also export DOCKER_HOST in your active shell
# socket path can be checked with `colima status`
export DOCKER_HOST=$HOME/.colima/default/docker.sock
```

If your primary need is communication across ports, then yes, anything available on the host is accessible via the `host.docker.internal` DNS. 
e.g. if the oracle container is available on port `1234` on the host, all other containers can reach it at `host.docker.internal:1234` irrespective of the Colima profile they are running inside of.

**Reply by wzxu** [reply]
Thanks. Didn't think about the possibility to start two Colima, and that I can switch to another while keep the original one still running. Will give this a try.

**Reply by wzxu** [reply]
Update: So this totally works. 
Database preparation is painfully slow, though (as expected). But smaller operations afterwards are *somewhat* acceptable.
[Keeping an eye on this discussion](https://github.com/oracle/docker-images/discussions/1951#discussioncomment-4880790) because it seems like it's actually possible to run x86_64 oracle binary on arm64 linux (or at least arm64 docker)?
Fingers crossed.

**Reply by rfay** [reply]
You can change between multiple running Colima instances using `docker context use <colimaprofile>`
